### PR TITLE
Backporting [SPARK-16494] Upgrade breeze version to 0.12

### PR DESCRIPTION
**Pull request by eugene-kharitonov**
## What changes were proposed in this pull request?

This pull request follows the pull request https://github.com/apache/spark/pull/14150 and updates the breeze version to 0.12. We are interested in it as it allows us to modify convergence criteria in LBFGS.

## How was this patch tested?

MLLib-specific unit tests pass.

### CODE REVIEW

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java [hunk 1]
```diff
@@ -108,7 +108,11 @@ public VectorPair call(Tuple2<Vector, Vector> pair) {
       .fit(df);
     List<Row> result = pca.transform(df).select("pca_features", "expected").toJavaRDD().collect();
     for (Row r : result) {
-      Assert.assertEquals(r.get(1), r.get(0));
```
Comment (superbobry) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java:111]: I wonder what was wrong with the previous version. This isn't breeze-version specific, is it?

File: dev/deps/spark-deps-hadoop-2.2 [hunk 2]
```diff
@@ -18,8 +18,8 @@ avro-ipc-1.7.7-tests.jar
 avro-ipc-1.7.7.jar
 avro-mapred-1.7.7-hadoop2.jar
 bonecp-0.8.0.RELEASE.jar
-breeze-macros_2.10-0.11.2.jar
-breeze_2.10-0.11.2.jar
+breeze-macros_2.11-0.12.jar
```
Comment (AnthonyTruchet) [inline dev/deps/spark-deps-hadoop-2.2:21]: we cannot switch to scala 2_11 I think (see #scala Slack channel for talk initiated by Sergeï)

File: dev/deps/spark-deps-hadoop-2.2 [hunk 3]
```diff
@@ -18,8 +18,8 @@ avro-ipc-1.7.7-tests.jar
...
+breeze_2.11-0.12.jar
```
Comment (AnthonyTruchet) [inline dev/deps/spark-deps-hadoop-2.2:22]: Idem: scala version

File: dev/deps/spark-deps-hadoop-2.2 [hunk 4]
```diff
@@ -170,6 +170,7 @@ scala-library-2.10.5.jar
 scala-reflect-2.10.5.jar
 scalap-2.10.5.jar
 servlet-api-2.5.jar
+shapeless_2.11-2.0.0.jar
```
Comment (AnthonyTruchet) [inline dev/deps/spark-deps-hadoop-2.2:173]: Idem

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java (hunk 1 above)
Comment (AnthonyTruchet) [reply mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java:111]: Probably that the previous version does not take numerical / distribution noie into account ?

**Review by AnthonyTruchet — CHANGES_REQUESTED** [review]

**Comment by AnthonyTruchet** [comment]
Looks like #22 just achieed the purpose of this request. I'm closing it.

**Pull request closed without merging**